### PR TITLE
feat(kubernetes): make reconcile support arrays

### LINKED ISSUE #110: Document expected structure of Jsonnet output
Hi team,

We're trying out tanka for managing our core kubernetes cluster configuration but have hit a few roadbumps.

Most of our problems appear to stem from lack of understanding about what type of object tanka is expecting your jsonnet to generate from the environments file.

From looking at your examples, it appears the implicit expectation is that tanka will recursively descend your jsonnet structure, assuming that your structure is a tree with kubernetes object leaf nodes. So that assumes that you have one global object which is a map of string to object, where that object is either a kubernetes resource or another tree.

```haskell
data Resource = Resource {
  apiVersion :: String, kind :: String, metadata :: Metadata
}
data Environment = Resource || [(string, Environment)]
```

If this is so, it would be great to document it. I suspect we might be missing something though, as tanka panics when we construct several otherwise valid configs, such as:

```
{
  a: {
    apiVersion: 'v1',
    kind: 'List',
    items: [/* valid k8s resources */],
  },
  b: self.a,
}
```

Lacking a clear explanation about what this format is makes it difficult to adopt tanka as a tool. All the advertising says tanka is a great `function`, but we don't know what the type signature is.

Can you help us out? Cheers.

### PR DESCRIPTION
**Pull request by sh0rez**
Adds array support for Reconcile.

Also documents the expected jsonnet structure. Fixes #110

### CODE REVIEW

**Comment by lawrencejones** [comment]
Hey @sh0rez, I think we have a branch that might do this, along with a few clean-ups we've made along the way. We'll push it in about an hour if you can wait for that!

I think your documentation is looking ace though 👌

**Comment by lawrencejones** [comment]
https://github.com/grafana/tanka/pull/109 is a separate take on this, hopefully with some simplifications that make tanka a bit easier to work on.

File: docs/jsonnet/expected-structure.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
+# Expected structure
+
+Tanka evaluates the `main.jsonnet` file of your [Environment](/environments) and
+filters the output (either Object or Array) for valid Kubernetes objects.  
+An object is considered valid if it has both, a `kind` and a `metadata.name` set.
```
[the remark this answers is not in the document]
Comment (rfratto) [reply docs/jsonnet/expected-structure.md]: I agree with @lawrencejones but also this remark seems inaccurate. `reconcile.go` seems to only be logging for `kind` and `apiVersion` right now. We should be checking for all three and document them here.

File: docs/jsonnet/expected-structure.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+## Deeply nested object (Recommended)
+The most commonly used structure is a single big object that includes all of
+your configs to be applied to the cluster nested under keys.  
+How deeply encapsulated the actual object is does not matter, Tanka will
+traverse down until it finds something that has both, a `kind` and a
+`metadata.name`.  
+
+??? Example
```
Comment (rfratto) [inline docs/jsonnet/expected-structure.md]: Should this be here?

File: docs/jsonnet/expected-structure.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        }
+      }
+    }
+    ```
+    
+Using this technique has the big benefit that it is self-documentary, as the
+nesting of keys can be used to logically group related manifests, for example by
+application.
+
+!!! info
```
Comment (rfratto) [inline docs/jsonnet/expected-structure.md]: Should this be here?

File: pkg/kubernetes/reconcile.go [hunk 4]
```diff
@@ -20,14 +20,37 @@ func (e ErrorPrimitiveReached) Error() string {
 }
 
 // walkJSON traverses deeply nested kubernetes manifest and extracts them into a flat []dict.
-func walkJSON(deep map[string]interface{}, path string) ([]map[string]interface{}, error) {
+func walkJSON(rawDeep interface{}, path string) ([]map[string]interface{}, error) {
+	flat := make([]map[string]interface{}, 0)
+
+	// array: walkJSON for each
+	if d, ok := rawDeep.([]map[string]interface{}); ok {
```
Comment (rfratto) [inline pkg/kubernetes/reconcile.go]: Just want to point out that this wouldn't support nested arrays, which might not be something we want t support right now anyway.

File: docs/jsonnet/expected-structure.md (hunk 2 above)
Comment (sh0rez) [reply docs/jsonnet/expected-structure.md]: Yeah, special markup from `mkdocs`, our docs generator

File: docs/jsonnet/expected-structure.md (hunk 3 above)
Comment (sh0rez) [reply docs/jsonnet/expected-structure.md]: see above

File: pkg/kubernetes/reconcile.go (hunk 4 above)
Comment (sh0rez) [reply pkg/kubernetes/reconcile.go]: I'm fine with this limitation, as the implementation would become much more complex, due to extensive type checking. 

Would like to encourage users to use dicts anyways

**Comment by sh0rez** [comment]
> I agree with @lawrencejones but also this remark seems inaccurate. reconcile.go seems to only be logging for kind and apiVersion right now. We should be checking for all three and document them here.

That's a good idea, just not in this particular PR. Will open an issue, let's address once #111 is done

**Pull request merged**